tests/visuals/test_visual_app.py

```diff
@@ -228,18 +231,33 @@ def test_missing_columns_validation(self) -> None:
                 selected_requests=SELECTED_REQUESTS_U2I,
             )
 
-        # Missing `Columns.Model` in reco pd.DataFrame
-        with pytest.raises(KeyError):
-            incorrect_reco = pd.DataFrame(
-                {Columns.User: [1, 2, 3, 4], Columns.Item: [3, 4, 3, 4], Columns.Score: [0.99, 0.9, 0.5, 0.5]}
-            )
-            AppDataStorage.from_raw(
-                reco=incorrect_reco,
-                item_data=ITEM_DATA,
-                interactions=INTERACTIONS,
-                is_u2i=True,
-                selected_requests=SELECTED_REQUESTS_U2I,
-            )
+    def test_successful_path_with_missing_model(self) -> None:
+        # Missing `Columns.Model`
+        reco_without_model = pd.DataFrame(
+            {Columns.User: [1, 2, 3, 4], Columns.Item: [3, 4, 3, 4], Columns.Score: [0.99, 0.9, 0.5, 0.5]}
+        )
+        ads = AppDataStorage.from_raw(
+            reco=reco_without_model,
+            item_data=ITEM_DATA,
+            interactions=INTERACTIONS,
+            is_u2i=True,
+            selected_requests=SELECTED_REQUESTS_U2I,
+        )
+        expected_grouped_reco = {
+            "model": {
+                "user_one": pd.DataFrame(
+                    {Columns.Item: [3], "feature_1": ["one"], Columns.Score: [0.99]}
+                ),
+                "user_three": pd.DataFrame(
+                    {Columns.Item: [3], "feature_1": ["one"], Columns.Score: [0.5]}
+                )
+            }
+        }
+        assert expected_grouped_reco.keys() == ads.grouped_reco.keys()
```

Comment:
Great to see this added to the test!

Could you please move this block of code (lines 256-26) to a separate method of the class? You can call it `assert_equal_data_in_grouped_reco_and_ads`.
Right now we are copy-pasting this line 3 times in the file and it's too much.